src/mantle/support/class-service-provider.php

```diff
@@ -133,4 +148,126 @@ public function add_command( $command ): Service_Provider {
 
 		return $this;
 	}
+
+	/**
+	 * Setup an after resolving listener, or fire immediately if already resolved.
+	 *
+	 * @param  string   $name Abstract name.
+	 * @param  callable $callback Callback.
+	 * @return void
+	 */
+	protected function call_after_resolving( string $name, callable $callback ): void {
+		$this->app->after_resolving( $name, $callback );
+
+		if ( $this->app->resolved( $name ) ) {
+			$callback( $this->app->make( $name ), $this->app );
+		}
+	}
+
+	/**
+	 * Register paths to be published by the publish command.
+	 *
+	 * @param string[]                  $paths Paths to publish.
+	 * @param string|array<string>|null $tags Tags to publish.
+	 */
+	public function publishes( array $paths, $tags = null ): void {
+		$class = static::class;
+
+		if ( ! array_key_exists( $class, static::$publishes ) ) {
+			static::$publishes[ $class ] = [];
+		}
+
+		static::$publishes[ $class ] = array_merge( static::$publishes[ $class ], $paths );
+
+		foreach ( (array) $tags as $tag ) {
+			if ( ! array_key_exists( $tag, static::$publish_tags ) ) {
+				static::$publish_tags[ $tag ] = [];
+			}
+
+			static::$publish_tags[ $tag ] = array_merge(
+				static::$publish_tags[ $tag ],
+				$paths,
+			);
+		}
+	}
+
+	/**
+	 * Get the service providers available for publishing.
+	 *
+	 * @return array<class-string<Service_Provider>>
+	 */
+	public static function publishable_providers() {
+		return array_keys( static::$publishes );
+	}
+
+	/**
+	 * Get the groups available for publishing.
+	 *
+	 * @return array<string>
```

Comment:
In some places you are using `string[]`. Suggesting it, for consistency.
```suggestion
	 * @return string[]
```